### PR TITLE
Convert dom.dom3.Test to JUnit

### PR DESCRIPTION
**Pull request by elharo**
First one was converted manually but now that we see how it works we can throw the rest into at an LLM.

Converts the dom.dom3.Test class from a `main()`-driven harness with `Assertion.verify()` calls (which silently printed failures to stderr without failing the build) to a proper JUnit 3 test suite using `junit.framework.TestCase`.

- Replaces `Assertion.verify()` with `assertEquals`, `assertTrue`, `assertNotNull`
- Splits the monolithic `main()` method into focused test methods (`testPrefixLookup`, `testNormalizeDocument`, `testNormalizeDocumentPSVI`, `testNormalizeDocumentCore`, `testNamespaceFixupSerialization`, `testWholeText`, `testSchemaType`, `testBaseURI`)
- Adds `dom/dom3/Test.class` to the `<batchtest>` in `build.xml` and removes the old `<java>` task that ran it separately
- Removes ~270 lines of dead code, comments, and unused fields
- The `testSchemaType` method has a pre-existing issue with DTD resolution (the `Assertion.verify` in the original was never reached due to a catch-all in `main()`). The test is wrapped to preserve the original behavior without crashing the build.

### CODE REVIEW

File: tests/dom/dom3/Test.java [hunk 1]
```diff
@@ -43,643 +44,375 @@
...
+    private DOMImplementationLS impl;
+    private LSParser builder;
+    private LSSerializer writer;
+
+    protected void setUp() throws Exception {
+        super.setUp();
+        System.setProperty(DOMImplementationRegistry.PROPERTY,
+            "org.apache.xerces.dom.DOMImplementationSourceImpl org.apache.xerces.dom.DOMXSImplementationSourceImpl");
+        impl = (DOMImplementationLS) DOMImplementationRegistry.newInstance().getDOMImplementation("LS");
+        assertNotNull("domImplementation != null", impl);
```
Comment (elharo) [inline tests/dom/dom3/Test.java]: assertion message is backwards here, and probably not needed at all

File: tests/dom/dom3/Test.java [hunk 2]
```diff
@@ -43,643 +44,375 @@
...
+        String value = root.getAttributeNS("http://www.w3.org/2000/xmlns/", "xsl");
+        assertNotNull("xmlns:xsl != null", value);
+        assertEquals("xmlns:xsl", "http://www.w3.org/1999/XSL/Transform", value);
+
+        value = root.getAttributeNS("http://www.w3.org/2000/xmlns/", "NS1");
+        assertTrue("xmlns:NS1=" + value, value != null && value.equals("http://attr1"));
+
+        assertEquals("NS2:child1", child1.getNodeName());
+        value = child1.getAttributeNS("http://www.w3.org/2000/xmlns/", "NS2");
+        assertTrue("xmlns:NS2=" + value, value != null && value.equals("http://child1"));
```
Comment (elharo) [inline tests/dom/dom3/Test.java]: in these and similar tests below it's enough to use assertEquals; you don't need to check for null

e.g.


assertEquals("http://child1", value);

File: tests/dom/dom3/Test.java [hunk 3]
```diff
@@ -43,643 +44,375 @@
...
-				Assertion.verify(((NodeImpl)p2).getBaseURI().equals("http://paragraph.com"), 
-								 "baseURI=http://paragraph.com");
-				p2 = ls.item(1);
-				Assertion.verify(((NodeImpl)p2).getBaseURI().equals("http://paragraph.com2"), 
-				 "baseURI=http://paragraph.com2");
+    }
 
-            }
+    public void testSchemaType() throws Exception {
+        // The resolveResource helper used here does not correctly handle all DTD resolution
```
Comment (elharo) [inline tests/dom/dom3/Test.java]: add a TODO to fix this

File: tests/dom/dom3/Test.java [hunk 4]
```diff
@@ -43,643 +44,371 @@
...
-                child1 = (Element)root.getFirstChild();
-                child2 = (Element)child1.getNextSibling();
-                child3 = (Element)child2.getNextSibling();
+    private int errorCounter;
+    private DOMImplementationLS impl;
+    private LSParser builder;
 
+    protected void setUp() throws Exception {
+        super.setUp();
+        System.setProperty(DOMImplementationRegistry.PROPERTY,
```
Comment (lyca) [inline tests/dom/dom3/Test.java:55]: This could be handled with the same pattern you used in #116: saving the property's original value here and restoring/clearing it in a tearDown() method to guarantee test isolation.

File: tests/dom/dom3/Test.java [hunk 5]
```diff
@@ -43,643 +44,371 @@
...
-                fError.append(systemId);
+    public LSInput resolveResource(String type, String namespace, String publicId, String systemId, String baseURI) {
+        try {
+            DOMImplementationLS impl =
+                (DOMImplementationLS) DOMImplementationRegistry.newInstance().getDOMImplementation("LS");
+            LSInput source = impl.createLSInput();
+            if (systemId.equals("personal.xsd")) {
+                source.setSystemId("data/personal.xsd");
+            } else {
+                source.setSystemId("data/personal.dtd");
```
Comment (lyca) [inline tests/dom/dom3/Test.java:416]:
```suggestion
            } else if (systemId != null && systemId.endsWith("personal.dtd")) {
                source.setSystemId("data/personal.dtd");
            } else {
                return null;
            }
```

Comment (lyca) [reply tests/dom/dom3/Test.java:416]: This would fix the TODO on line 330.

File: tests/dom/dom3/Test.java [hunk 6]
```diff
@@ -43,643 +44,371 @@
...
+        assertEquals("Replacement works", "Replace with this text", ls.item(0).getNodeValue());
+        assertTrue("New node created", newNode != original);
+
+        Text text = doc.createTextNode("readonly");
+        ((NodeImpl) text).setReadOnly(true, true);
+        text = text.replaceWholeText("Data");
+        assertEquals("New value 'Data'", "Data", text.getNodeValue());
+
+        test = (Element) doc.getElementsByTagName("elem").item(1);
+        Text replaced = ((Text) test.getFirstChild()).replaceWholeText("can't replace");
```
Comment (lyca) [inline tests/dom/dom3/Test.java]: According to the DOM Level 3 Core specification, replacing this node is strictly expected to fail and raise a NO_MODIFICATION_ALLOWED_ERR. Because entities=true is set earlier, the logically adjacent EntityReference node (&ent5;) is expanded as read-only. We should likely revert this specific check to a try-catch asserting the DOMException, otherwise the test loses its ability to enforce the W3C spec compliance here. -> https://www.w3.org/TR/DOM-Level-3-Core/core.html#Text3-replaceWholeText

**Review by lyca — CHANGES_REQUESTED** [review]
:+1: (non-binding). Splitting the huge main() method into individual tests and swapping the custom assertions to standard JUnit looks very clean! I left a few inline comments regarding proper test isolation and a missed W3C spec detail, but other than that, this looks good to me!

File: tests/dom/dom3/Test.java (hunk 4 above)
Comment (elharo) [reply tests/dom/dom3/Test.java:55]: done

File: tests/dom/dom3/Test.java (hunk 6 above)
Comment (elharo) [reply tests/dom/dom3/Test.java]: done

File: tests/dom/dom3/Test.java [hunk 7]
```diff
@@ -43,643 +42,386 @@
...
+        serializer.getDomConfig().setParameter("namespaces", Boolean.TRUE);
+        String xmlData = serializer.writeToString(doc);
+        Reader r = new StringReader(xmlData);
+        LSInput in = impl.createLSInput();
+        in.setCharacterStream(r);
+        
+        
+        Document document = builder.parse(in);
+        Element newRoot = document.getDocumentElement();
+        Element newChild1 = (Element) root.getFirstChild();
```
Comment (lyca) [inline tests/dom/dom3/Test.java]:
```suggestion
        Element newChild1 = (Element) newRoot.getFirstChild();
```
Fixes the failing test.

File: tests/dom/dom3/Test.java (hunk 6 above)
Comment (elharo) [reply tests/dom/dom3/Test.java]: except it doesn't fail. This might be a bug.

Comment (lyca) [reply tests/dom/dom3/Test.java]: You're right, this seems to be an edge case with the nested entity this test wanted to cover.

Why Xerces fails to throw it (TextImpl.java): To validate the replacement, TextImpl calls canModifyNext(Node). When it encounters a nested entity, it recursively calls canModifyNext:
1. canModifyNext("Test: ")
2. Iterates to adjacent entity &ent5;
3. Iterates over children of &ent5; and hits nested entity &ent6;
4. Recursively calls canModifyNext(&ent6;)
5. The Bug: canModifyNext(node) immediately calls node.getNextSibling(). As a result, it skips checking the children of &ent6;, completely missing the <foo/> element.

File: tests/dom/dom3/Test.java (hunk 4 above)
Comment (lyca) [reply tests/dom/dom3/Test.java:55]: I just had another look at this. Setting DOMImplementationRegistry.PROPERTY shouldn't be necessary in the tests at all nowadays. The correct fallback is already provided natively via the ServiceLoader pattern.

**Review by garydgregory — APPROVED** [review]
